tests/unit/s2n_client_early_data_indication_test.c

```diff
@@ -297,5 +297,118 @@ int main(int argc, char **argv)
         }
     }
 
+    /* Test state transitions with a HelloRetryRequest.
+     *
+     *= https://tools.ietf.org/rfc/rfc8446#section-4.2.10
+     *= type=test
+     *# A server which receives an "early_data" extension MUST behave in one
+     *# of three ways:
+     *
+     *= https://tools.ietf.org/rfc/rfc8446#section-4.2.10
+     *= type=test
+     *# -  Request that the client send another ClientHello by responding
+     *#    with a HelloRetryRequest.
+     */
+    {
+        /* Hello Retry Request because of rejected early data.
+         *
+         * The S2N server does not reject early data via a HelloRetryRequest, but other implementations might.
```

Comment:
In `s2n_set_hello_retry_required` below it looks like we do reject early data when mark set HRR required. Am I following this correctly?

Reply:
A client requests early data support by sending the early_data_indication extension. The server can reject that request either by not including the early_data_indication extension on its ServerHello and proceeding with the handshake as if the client had not sent the extension, or by responding with a HRR.

The S2N server does the former, so it only initiates a HRR in response to a missing keyshare. But since other server implementations might do the later, we need to test that the S2N client can handle a server sending a HRR just to reject early data and not to handle a missing keyshare. This test forces the S2N server to retry when it otherwise wouldn't by calling s2n_set_hello_retry_required directly, then verifies that the client handles that HRR properly.